Samples: Blob.from_string() is deprecated. Use Blob.from_uri() instead.

See the samples test failure in PR https://github.com/googleapis/python-documentai-toolbox/pull/380

https://github.com/googleapis/python-documentai-toolbox/blob/0e59848a43b486fd61e022e48f08b93104076eb0/google/cloud/documentai_toolbox/utilities/gcs_utilities.py#L145

https://btx.cloud.google.com/invocations/41ddcb73-85dd-4eda-80d5-052e7c29805a/log

```
=================================== FAILURES ===================================
________________________ test_quickstart_sample_gcs_uri ________________________

capsys = <_pytest.capture.CaptureFixture object at 0x7f9fb13b9880>

    def test_quickstart_sample_gcs_uri(capsys: pytest.CaptureFixture) -> None:
        gcs_uri = (
            "gs://documentai_toolbox_samples/output/123456789/0/toolbox_invoice_test-0.json"
        )
>       quickstart_sample.quickstart_sample(gcs_uri=gcs_uri)

test_quickstart_sample.py:46: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
quickstart_sample.py:68: in quickstart_sample
    wrapped_document = document.Document.from_gcs_uri(gcs_uri=gcs_uri)
../../google/cloud/documentai_toolbox/wrappers/document.py:566: in from_gcs_uri
    blob = gcs_utilities.get_blob(gcs_uri=gcs_uri, module="get-document")
../../google/cloud/documentai_toolbox/utilities/gcs_utilities.py:145: in get_blob
    return storage.Blob.from_string(gcs_uri, _get_storage_client(module=module))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = 
uri = 'gs://documentai_toolbox_samples/output/123456789/0/toolbox_invoice_test-0.json'
client = 

    @classmethod
    def from_string(cls, uri, client=None):
        """(Deprecated) Get a constructor for blob object by URI.
    
        .. note::
           Deprecated alias for :meth:`from_uri`.
    
        .. code-block:: python
    
            from google.cloud import storage
            from google.cloud.storage.blob import Blob
            client = storage.Client()
            blob = Blob.from_string("gs://bucket/object", client=client)
    
        :type uri: str
        :param uri: The blob uri following a gs://bucket/object pattern.
          Both a bucket and object name is required to construct a blob object.
    
        :type client: :class:`~google.cloud.storage.client.Client`
        :param client:
            (Optional) The client to use.  Application code should
            *always* pass ``client``.
    
        :rtype: :class:`google.cloud.storage.blob.Blob`
        :returns: The blob object created.
        """
>       warnings.warn(_FROM_STRING_DEPRECATED, PendingDeprecationWarning, stacklevel=2)
E       PendingDeprecationWarning: Blob.from_string() is deprecated and will be removed in future. Use Blob.from_uri() instead.

.nox/py-3-8/lib/python3.8/site-packages/google/cloud/storage/blob.py:449: PendingDeprecationWarning
-------- generated xml file: /workspace/samples/snippets/sponge_log.xml --------
```

--- a/.coveragerc
+++ b/.coveragerc
@@ -22,7 +22,7 @@ omit =
   google/cloud/__init__.py
 
 [report]
-fail_under = 98
+fail_under = 99
 show_missing = True
 exclude_lines =
     # Re-enable the standard pragma

--- a/.coveragerc
+++ b/.coveragerc
@@ -22,7 +22,7 @@ omit =
   google/cloud/__init__.py
 
 [report]
-fail_under = 99
+fail_under = 98
 show_missing = True
 exclude_lines =
     # Re-enable the standard pragma

--- a/.github/workflows/unittest.yml
+++ b/.github/workflows/unittest.yml
@@ -58,4 +58,4 @@ jobs:
       run: |
         find .coverage-results -type f -name '*.zip' -exec unzip {} \;
         coverage combine .coverage-results/**/.coverage*
-        coverage report --show-missing --fail-under=99
+        coverage report --show-missing --fail-under=98

--- a/google/cloud/documentai_toolbox/utilities/gcs_utilities.py
+++ b/google/cloud/documentai_toolbox/utilities/gcs_utilities.py
@@ -19,7 +19,7 @@ import re
 from typing import Dict, List, Optional, Tuple
 
 from google.api_core.gapic_v1 import client_info
-import pkg_resources
+import importlib.metadata
 
 from google.cloud import documentai, documentai_toolbox, storage
 from google.cloud.documentai_toolbox import constants

--- a/google/cloud/documentai_toolbox/utilities/gcs_utilities.py
+++ b/google/cloud/documentai_toolbox/utilities/gcs_utilities.py
@@ -14,12 +14,12 @@
 # limitations under the License.
 #
 """Google Cloud Storage utilities."""
+import importlib.metadata
 import os
 import re
 from typing import Dict, List, Optional, Tuple
 
 from google.api_core.gapic_v1 import client_info
-import importlib.metadata
 
 from google.cloud import documentai, documentai_toolbox, storage
 from google.cloud.documentai_toolbox import constants

--- a/google/cloud/documentai_toolbox/utilities/gcs_utilities.py
+++ b/google/cloud/documentai_toolbox/utilities/gcs_utilities.py
@@ -142,7 +142,7 @@ def get_blob(
     if not re.match(constants.FILE_CHECK_REGEX, gcs_uri):
         raise ValueError("gcs_uri must link to a single file.")
 
-    return storage.Blob.from_string(gcs_uri, _get_storage_client(module=module))
+    return storage.Blob.from_uri(gcs_uri, _get_storage_client(module=module))
 
 
 def split_gcs_uri(gcs_uri: str) -> Tuple[str, str]:

--- a/google/cloud/documentai_toolbox/utilities/gcs_utilities.py
+++ b/google/cloud/documentai_toolbox/utilities/gcs_utilities.py
@@ -15,6 +15,7 @@
 #
 """Google Cloud Storage utilities."""
 import os
+import pkg_resources
 import re
 from typing import Dict, List, Optional, Tuple
 
@@ -142,7 +143,18 @@ def get_blob(
     if not re.match(constants.FILE_CHECK_REGEX, gcs_uri):
         raise ValueError("gcs_uri must link to a single file.")
 
-    return storage.Blob.from_uri(gcs_uri, _get_storage_client(module=module))
+    try:
+        version = pkg_resources.get_distribution("google-cloud-storage").version
+    except pkg_resources.DistributionNotFound:
+        raise ImportError("google-cloud-storage is not installed.")
+
+    client = _get_storage_client(module=module)
+
+    major, _, _ = map(int, version.split("."))
+    if major < 3:
+        return storage.Blob.from_string(gcs_uri, client)
+    else:
+        return storage.Blob.from_uri(gcs_uri, client)
 
 
 def split_gcs_uri(gcs_uri: str) -> Tuple[str, str]:

--- a/google/cloud/documentai_toolbox/utilities/gcs_utilities.py
+++ b/google/cloud/documentai_toolbox/utilities/gcs_utilities.py
@@ -15,11 +15,11 @@
 #
 """Google Cloud Storage utilities."""
 import os
-import pkg_resources
 import re
 from typing import Dict, List, Optional, Tuple
 
 from google.api_core.gapic_v1 import client_info
+import pkg_resources
 
 from google.cloud import documentai, documentai_toolbox, storage
 from google.cloud.documentai_toolbox import constants

--- a/google/cloud/documentai_toolbox/utilities/gcs_utilities.py
+++ b/google/cloud/documentai_toolbox/utilities/gcs_utilities.py
@@ -144,8 +144,8 @@ def get_blob(
         raise ValueError("gcs_uri must link to a single file.")
 
     try:
-        version = pkg_resources.get_distribution("google-cloud-storage").version
-    except pkg_resources.DistributionNotFound:
+        version = importlib.metadata.version("google-cloud-storage")
+    except importlib.metadata.PackageNotFoundError:
         raise ImportError("google-cloud-storage is not installed.")
 
     client = _get_storage_client(module=module)

--- a/google/cloud/documentai_toolbox/utilities/gcs_utilities.py
+++ b/google/cloud/documentai_toolbox/utilities/gcs_utilities.py
@@ -14,7 +14,6 @@
 # limitations under the License.
 #
 """Google Cloud Storage utilities."""
-import importlib.metadata
 import os
 import re
 from typing import Dict, List, Optional, Tuple
@@ -143,18 +142,11 @@ def get_blob(
     if not re.match(constants.FILE_CHECK_REGEX, gcs_uri):
         raise ValueError("gcs_uri must link to a single file.")
 
-    try:
-        version = importlib.metadata.version("google-cloud-storage")
-    except importlib.metadata.PackageNotFoundError:
-        raise ImportError("google-cloud-storage is not installed.")
+    # google-cloud-storage >= 3.0.0
+    if hasattr(storage.Blob, "from_uri"):
+        return storage.Blob.from_uri(gcs_uri, _get_storage_client(module=module))
 
-    client = _get_storage_client(module=module)
-
-    major, _, _ = map(int, version.split("."))
-    if major < 3:
-        return storage.Blob.from_string(gcs_uri, client)
-    else:
-        return storage.Blob.from_uri(gcs_uri, client)
+    return storage.Blob.from_string(gcs_uri, _get_storage_client(module=module))
 
 
 def split_gcs_uri(gcs_uri: str) -> Tuple[str, str]:

--- a/owlbot.py
+++ b/owlbot.py
@@ -30,7 +30,7 @@ templated_files = common.py_library(
     unit_test_python_versions=["3.8", "3.9", "3.10", "3.11", "3.12", "3.13", "3.14"],
     system_test_python_versions=["3.9", "3.14"],
     default_python_version="3.13",
-    cov_level=99,
+    cov_level=98,
     intersphinx_dependencies={
         "pandas": "https://pandas.pydata.org/pandas-docs/stable/"
     },

--- a/tests/unit/test_gcs_utilities.py
+++ b/tests/unit/test_gcs_utilities.py
@@ -13,7 +13,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import importlib
 from unittest import mock
 
 import pytest

--- a/tests/unit/test_gcs_utilities.py
+++ b/tests/unit/test_gcs_utilities.py
@@ -13,6 +13,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import importlib
 from unittest import mock
 
 import pytest
@@ -579,3 +580,90 @@ def test_get_blobs_with_no_input():
         match="You must provide either `gcs_uri` or both `gcs_bucket_name` and `gcs_prefix`.",
     ):
         gcs_utilities.get_blobs()
+
+
+@mock.patch("google.cloud.documentai_toolbox.utilities.gcs_utilities.storage")
+def test_get_blobs_with_gcs_uri(mock_storage):
+    client = mock_storage.Client.return_value
+    gcs_uri = "gs://test-bucket/test-directory/1/"
+
+    gcs_utilities.get_blobs(gcs_uri=gcs_uri)
+
+    mock_storage.Client.assert_called_once()
+    client.list_blobs.assert_called_once_with("test-bucket", prefix="test-directory/1/")
+
+
+def test_get_blobs_with_file_type_error():
+    with pytest.raises(ValueError, match="gcs_prefix cannot contain file types"):
+        gcs_utilities.get_blobs(gcs_bucket_name="test-bucket", gcs_prefix="test.json")
+
+
+@mock.patch("google.cloud.documentai_toolbox.utilities.gcs_utilities.storage")
+def test_get_blob_success_major_3(mock_storage):
+    mock_version = "3.0.0"
+    with mock.patch("importlib.metadata.version", return_value=mock_version):
+        client = mock_storage.Client.return_value
+        gcs_uri = "gs://test-bucket/test.json"
+
+        gcs_utilities.get_blob(gcs_uri)
+
+        mock_storage.Blob.from_uri.assert_called_once_with(gcs_uri, client)
+
+
+@mock.patch("google.cloud.documentai_toolbox.utilities.gcs_utilities.storage")
+def test_get_blob_success_major_2(mock_storage):
+    mock_version = "2.0.0"
+    with mock.patch("importlib.metadata.version", return_value=mock_version):
+        client = mock_storage.Client.return_value
+        gcs_uri = "gs://test-bucket/test.json"
+
+        gcs_utilities.get_blob(gcs_uri)
+
+        mock_storage.Blob.from_string.assert_called_once_with(gcs_uri, client)
+
+
+def test_get_blob_invalid_uri():
+    with pytest.raises(ValueError, match="gcs_uri must link to a single file."):
+        gcs_utilities.get_blob("gs://test-bucket/prefix/")
+
+
+def test_get_blob_import_error():
+    with mock.patch(
+        "importlib.metadata.version",
+        side_effect=importlib.metadata.PackageNotFoundError,
+    ):
+        with pytest.raises(ImportError, match="google-cloud-storage is not installed."):
+            gcs_utilities.get_blob("gs://test-bucket/test.json")
+
+
+@mock.patch("google.cloud.documentai_toolbox.utilities.gcs_utilities.storage")
+def test_print_gcs_document_tree_with_skipping_files(mock_storage, capfd):
+    client = mock_storage.Client.return_value
+    mock_bucket = mock.Mock()
+    client.Bucket.return_value = mock_bucket
+
+    blobs = [
+        storage.Blob(
+            name=f"gs://test-directory/1/test_shard{i}.json",
+            bucket="gs://test-directory/1",
+        )
+        for i in range(1, 11)
+    ]
+
+    client.list_blobs.return_value = blobs
+
+    # files_to_display = 2. 10 files total.
+    # idx 0, 1, 2 -> print
+    # idx 3, 4, 5, 6, 7, 8 -> skip
+    # idx 9 -> print last
+    gcs_utilities.print_gcs_document_tree(
+        gcs_bucket_name="test-directory", gcs_prefix="/", files_to_display=2
+    )
+
+    out, err = capfd.readouterr()
+    assert "├──test_shard1.json" in out
+    assert "├──test_shard2.json" in out
+    assert "├──test_shard3.json" in out
+    assert "├──test_shard4.json" not in out
+    assert "│  ...." in out
+    assert "└──test_shard10.json" in out

--- a/tests/unit/test_gcs_utilities.py
+++ b/tests/unit/test_gcs_utilities.py
@@ -598,72 +598,32 @@ def test_get_blobs_with_file_type_error():
         gcs_utilities.get_blobs(gcs_bucket_name="test-bucket", gcs_prefix="test.json")
 
 
-@mock.patch("google.cloud.documentai_toolbox.utilities.gcs_utilities.storage")
-def test_get_blob_success_major_3(mock_storage):
-    mock_version = "3.0.0"
-    with mock.patch("importlib.metadata.version", return_value=mock_version):
-        client = mock_storage.Client.return_value
-        gcs_uri = "gs://test-bucket/test.json"
-
-        gcs_utilities.get_blob(gcs_uri)
-
-        mock_storage.Blob.from_uri.assert_called_once_with(gcs_uri, client)
-
-
-@mock.patch("google.cloud.documentai_toolbox.utilities.gcs_utilities.storage")
-def test_get_blob_success_major_2(mock_storage):
-    mock_version = "2.0.0"
-    with mock.patch("importlib.metadata.version", return_value=mock_version):
-        client = mock_storage.Client.return_value
-        gcs_uri = "gs://test-bucket/test.json"
-
-        gcs_utilities.get_blob(gcs_uri)
-
-        mock_storage.Blob.from_string.assert_called_once_with(gcs_uri, client)
-
-
 def test_get_blob_invalid_uri():
     with pytest.raises(ValueError, match="gcs_uri must link to a single file."):
         gcs_utilities.get_blob("gs://test-bucket/prefix/")
 
 
-def test_get_blob_import_error():
-    with mock.patch(
-        "importlib.metadata.version",
-        side_effect=importlib.metadata.PackageNotFoundError,
-    ):
-        with pytest.raises(ImportError, match="google-cloud-storage is not installed."):
-            gcs_utilities.get_blob("gs://test-bucket/test.json")
+@mock.patch("google.cloud.documentai_toolbox.utilities.gcs_utilities.storage")
+def test_get_blob_from_uri(mock_storage):
+    gcs_uri = "gs://test-bucket/test.json"
+
+    # Mock storage.Blob.from_uri to exist
+    mock_storage.Blob.from_uri.return_value = mock.Mock(spec=storage.blob.Blob)
+
+    gcs_utilities.get_blob(gcs_uri=gcs_uri)
+
+    mock_storage.Blob.from_uri.assert_called_once()
 
 
 @mock.patch("google.cloud.documentai_toolbox.utilities.gcs_utilities.storage")
-def test_print_gcs_document_tree_with_skipping_files(mock_storage, capfd):
-    client = mock_storage.Client.return_value
-    mock_bucket = mock.Mock()
-    client.Bucket.return_value = mock_bucket
+def test_get_blob_from_string(mock_storage):
+    gcs_uri = "gs://test-bucket/test.json"
 
-    blobs = [
-        storage.Blob(
-            name=f"gs://test-directory/1/test_shard{i}.json",
-            bucket="gs://test-directory/1",
-        )
-        for i in range(1, 11)
-    ]
+    # Mock storage.Blob to NOT have from_uri
+    del mock_storage.Blob.from_uri
 
-    client.list_blobs.return_value = blobs
+    mock_storage.Blob.from_string.return_value = mock.Mock(spec=storage.blob.Blob)
 
-    # files_to_display = 2. 10 files total.
-    # idx 0, 1, 2 -> print
-    # idx 3, 4, 5, 6, 7, 8 -> skip
-    # idx 9 -> print last
-    gcs_utilities.print_gcs_document_tree(
-        gcs_bucket_name="test-directory", gcs_prefix="/", files_to_display=2
-    )
+    gcs_utilities.get_blob(gcs_uri=gcs_uri)
 
-    out, err = capfd.readouterr()
-    assert "├──test_shard1.json" in out
-    assert "├──test_shard2.json" in out
-    assert "├──test_shard3.json" in out
-    assert "├──test_shard4.json" not in out
-    assert "│  ...." in out
-    assert "└──test_shard10.json" in out
+    mock_storage.Blob.from_string.assert_called_once()

